File: Segment_Tree_Query.cpp
/**
 * Definition of SegmentTreeNode:
 * class SegmentTreeNode {
 * public:
 *     int start, end, max;
 *     SegmentTreeNode *left, *right;
 *     SegmentTreeNode(int start, int end, int max) {
 *         this->start = start;
 *         this->end = end;
 *         this->max = max;
 *         this->left = this->right = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     *@param root, start, end: The root of segment tree and 
     *                         an segment / interval
     *@return: The maximum number in the interval [start, end]
     */
    int query(SegmentTreeNode *root, int start, int end) {
        // write your code here
        // if(root->start == start && root->end == end){
        //     return root->max;
        // }
        // SegmentTreeNode* left = root->left;
        // SegmentTreeNode* right = root->right;
        // if(end < right->start){
        //     return query(left, start, end);
        // }else if(start > left->end){
        //     return query(right, start, end);
        // }else{
        //     return max(query(left, start, left->end), query(right, right->start, end));
        // }
        //good codes
        if(root == nullptr || root->end < start || root->start > end){
            return INT_MIN;
        }
        if(root->end <= end && root->start >= start){
            return root->max;
        }
        int left = query(root->left, start, end);
        int right = query(root->right, start, end);
        return max(left, right);
    }
};

File: Segment Tree Query.py
"""
Definition of SegmentTreeNode:
class SegmentTreeNode:
    def __init__(self, start, end, max):
        self.start, self.end, self.max = start, end, max
        self.left, self.right = None, None
"""

class Solution:	
    # @param root, start, end: The root of segment tree and 
    #                          an segment / interval
    # @return: The maximum number in the interval [start, end]
    def query(self, root, start, end):
        # write your code here
        
        if root.start > end or root.end < start:
            return float('-inf')
        else:
            if root.start >= start and root.end <= end:
                return root.max
            else:
                return max(self.query(root.left, start, end), self.query(root.right, start, end))